streams/src/main/java/org/apache/kafka/streams/kstream/internals/GroupedStreamAggregateBuilder.java

```diff
@@ -85,18 +71,15 @@ public V apply() {
         OptimizableRepartitionNode.OptimizableRepartitionNodeBuilder<K, V> repartitionNodeBuilder = OptimizableRepartitionNode.optimizableRepartitionNodeBuilder();
 
         final String sourceName = repartitionIfRequired(storeBuilder.name(), repartitionNodeBuilder);
-        builder.internalTopologyBuilder.addProcessor(aggFunctionName, aggregateSupplier, sourceName);
-        builder.internalTopologyBuilder.addStateStore(storeBuilder, aggFunctionName);
-
 
         StreamsGraphNode parentNode = streamsGraphNode;
 
         if (!sourceName.equals(this.name)) {
             StreamsGraphNode repartitionNode = repartitionNodeBuilder.build();
             streamsGraphNode.addChildNode(repartitionNode);
+            builder.maybeAddNodeForOptimizationMetadata(repartitionNode);
```

Comment:
+1

Reply:
This is intentional. 
Today, when a repartition is required, the `InternalTopologyBuilder` creates a repartition operation and immediately writes it to the physical plan.  So we need to capture the repartition as a graph node and pass the new node to the `InternalStreamsBuilder` for possible metadata collection for optimization.  

In other cased graph nodes are created in classes that don't subclass the `AbstractStream`.

Reply:
Could you explain to us a bit more? I'm still scratching my head now on when do we call 

`AbstractStream#addGraphNode` 

v.s. 

`parentNode.addChildNode; builder.maybeAddNodeForOptimizationMetadata(repartitionNode);`

Because their logic are just the same, right?

Reply:
Yes, the logic is the same.  But `GroupedStreamAggregateBuilder` does not subclass `AbstractStream` so we need to make those two calls separately.

Secondly, in the example above we are building 2 graph nodes.  The repartition node and the stateful processor node for the aggregation.  

Today when we need to repartition for an aggregation, the `InternalTopologyBuilder` creates a repartition operation and it becomes the parent of the aggregation operation.  So the repartition node needs to be a child of the current `StreamsGraphNode` in the `GroupedStreamAggregateBuilder` but it must be the parent of the aggregation graph node. 

Does this make sense?